File: packages/android-app/app/src/main/java/com/lotus/android/sync/ThreeWayMerge.kt
package com.lotus.android.sync

import com.lotus.android.core.model.Vault
import com.lotus.android.core.model.VaultEntry

data class MergeConflict(
  val entryId: String,
  val base: VaultEntry?,
  val local: VaultEntry?,
  val remote: VaultEntry?
)

data class MergeResult(
  val merged: Vault,
  val conflicts: List<MergeConflict>
)

fun threeWayMerge(base: Vault, local: Vault, remote: Vault): MergeResult {
  val baseMap = base.entries.associateBy { it.id }
  val localMap = local.entries.associateBy { it.id }
  val remoteMap = remote.entries.associateBy { it.id }
  val allIds = (baseMap.keys + localMap.keys + remoteMap.keys).toSet()

  val mergedEntries = mutableListOf<VaultEntry>()
  val conflicts = mutableListOf<MergeConflict>()

  for (id in allIds) {
    val b = baseMap[id]
    val l = localMap[id]
    val r = remoteMap[id]
    when {
      l == r -> if (l != null) mergedEntries += l
      l == b -> if (r != null) mergedEntries += r
      r == b -> if (l != null) mergedEntries += l
      else -> {
        conflicts += MergeConflict(id, b, l, r)
        val best = listOfNotNull(l, r).maxByOrNull { it.modified }
        if (best != null) mergedEntries += best
      }
    }
  }

  return MergeResult(
    merged = local.copy(
      entries = mergedEntries.sortedByDescending { it.modified },
      syncVersion = maxOf(local.syncVersion, remote.syncVersion) + 1
    ),
    conflicts = conflicts
  )
}

File: packages/extension/src/lib/three-way-merge.ts
import type { Vault, VaultEntry } from '@lotus/shared'

export interface Conflict {
  entryId: string
  localEntry: VaultEntry | null
  remoteEntry: VaultEntry | null
  baseEntry: VaultEntry | null
}

export interface MergeResult {
  vault: Vault
  conflicts: Conflict[]
}

function entriesEqual(a: VaultEntry, b: VaultEntry): boolean {
  return JSON.stringify(a) === JSON.stringify(b)
}

export async function threeWayMerge(
  local: Vault,
  remote: Vault,
  base: Vault,
  constantTimeDelayMs = 500
): Promise<MergeResult> {
  const startTime = performance.now()

  const conflicts: Conflict[] = []
  const mergedEntries: VaultEntry[] = []
  const allEntryIds = new Set([
    ...local.entries.map((e) => e.id),
    ...remote.entries.map((e) => e.id),
    ...base.entries.map((e) => e.id)
  ])

  for (const entryId of allEntryIds) {
    const localEntry = local.entries.find((e) => e.id === entryId) || null
    const remoteEntry = remote.entries.find((e) => e.id === entryId) || null
    const baseEntry = base.entries.find((e) => e.id === entryId) || null

    if (!localEntry && !remoteEntry) continue
    if (!localEntry) {
      if (baseEntry) {
        conflicts.push({ entryId, localEntry, remoteEntry, baseEntry })
        // Preserve remote when local deleted but remote still present.
        if (remoteEntry) mergedEntries.push(remoteEntry)
      } else if (remoteEntry) {
        mergedEntries.push(remoteEntry)
      }
      continue
    }
    if (!remoteEntry) {
      if (baseEntry) {
        conflicts.push({ entryId, localEntry, remoteEntry, baseEntry })
        // Preserve local when remote deleted but local still present.
        mergedEntries.push(localEntry)
      } else {
        mergedEntries.push(localEntry)
      }
      continue
    }
    if (baseEntry && entriesEqual(localEntry, baseEntry)) mergedEntries.push(remoteEntry)
    else if (baseEntry && entriesEqual(remoteEntry, baseEntry)) mergedEntries.push(localEntry)
    else if (entriesEqual(localEntry, remoteEntry)) mergedEntries.push(localEntry)
    else {
      conflicts.push({ entryId, localEntry, remoteEntry, baseEntry })
      mergedEntries.push(localEntry)
    }
  }

  const result: MergeResult = {
    vault: {
      ...local,
      entries: mergedEntries,
      syncVersion: Math.max(local.syncVersion, remote.syncVersion) + 1,
      lastSync: Date.now()
    },
    conflicts
  }

  const elapsed = performance.now() - startTime
  const remaining = Math.max(0, constantTimeDelayMs - elapsed)

  if (remaining > 0) {
    await new Promise(resolve => setTimeout(resolve, remaining))
  }

  return result
}
